app/services/archetypeCacheService.js

```diff
@@ -1,106 +1,82 @@
-
-    var isDatatypeLookupLoading = false;
     var datatypeCache = [];
-
+    
     return {
-    	getDataTypeFromCache: function(guid) {
-        	return _.find(datatypeCache, function (dt){
-	            return dt.dataTypeGuid == guid;
-	        });
-    	},
-
-    	addDatatypeToCache: function(datatype, dataTypeGuid) {
-            var cachedDatatype = this.getDataTypeFromCache(dataTypeGuid);
-
-            if(!cachedDatatype) {
-            	datatype.dataTypeGuid = dataTypeGuid;
-            	datatypeCache.push(datatype);
-            }
-    	},
+        initialize: function() {
+            return archetypePropertyEditorResource.getAllDataTypesForCache().then(function(data) {
+                datatypeCache = data;
+            });
+        },
+        
+        getDataTypeFromCache: function(guid) {
+            return _.find(datatypeCache, function (dt){
+                return dt.dataTypeGuid == guid;
+            });
+        },
  
-		getDatatypeByGuid: function(guid) {
-			var cachedDatatype = this.getDataTypeFromCache(guid);
-
-	        if(cachedDatatype) {
-	            return cachedDatatype;
-	        }
-
-	        //go get it from server, but this should already be pre-populated from the directive, but I suppose I'll leave this in in case used ad-hoc
-	        if (!isDatatypeLookupLoading) {
-	            isDatatypeLookupLoading = true;
-
-	            archetypePropertyEditorResource.getDataType(guid).then(function(datatype) {
-
-	            	datatype.dataTypeGuid = guid;
-
-	                datatypeCache.push(datatype);
-
-	                isDatatypeLookupLoading = false;
-
-	                return datatype;
-	            });
-	        }
+        getDatatypeByGuid: function(guid) {
+            var cachedDatatype = this.getDataTypeFromCache(guid);
+            
+            if(cachedDatatype) {
+                return cachedDatatype;
+            }
 
-	        return null;
+            return null;
         },
 
-     	getEntityById: function(scope, id, type) {
-	        var cachedEntity = _.find(entityCache, function (e){
-	            return e.id == id;
-	        });
-
-	        if(cachedEntity) {
-	            return cachedEntity;
-	        }
-
-	        //go get it from server
-	        if (!isEntityLookupLoading) {
-	            isEntityLookupLoading = true;
-
-	            scope.resources.entityResource.getById(id, type).then(function(entity) {
-
-	                entityCache.push(entity);
-
-	                isEntityLookupLoading = false;
-
-	                return entity;
-	            });
-	        }
-
-	        return null;
-     	},
-
-     	getEntityByUmbracoId: function(scope, udi, type) {
-	        var cachedEntity = _.find(entityCache, function (e){
-	            return e.udi == udi;
-	        });
+        //perhaps this should return a promise?
+        getEntityById: function(id, type) {
+            var deferred = $q.defer();
+            
+            var cachedEntity = _.find(entityCache, function (e){
```

Comment:
Why not make `entityCache` an associative array (with the ID as key) rather than a plain array? Could save some CPU cycles by performing instant lookup rather than scanning every item in the array for a match.

Reply:
Great feedback, sounds like a quick win. 👍